src/pools/abstract-pool.ts

```diff
@@ -0,0 +1,173 @@
+
+  protected id: number = 0
+
+  public constructor (
+    public readonly numWorkers: number,
+    public readonly filePath: string,
+    public readonly opts: PoolOptions<Worker> = { maxTasks: 1000 }
+  ) {
+    if (!this.isMain()) {
+      throw new Error('Cannot start a pool from a worker!')
+    }
+    // TODO christopher 2021-02-07: Improve this check e.g. with a pattern or blank check
+    if (!this.filePath) {
+      throw new Error('Please specify a file with a worker implementation')
+    }
+
+    this.setupHook()
+
+    for (let i = 1; i <= this.numWorkers; i++) {
+      this.internalNewWorker()
+    }
+
+    this.emitter = new PoolEmitter()
+  }
+
+  protected setupHook (): void {
+    // Can be overridden
+  }
+
+  protected abstract isMain (): boolean
+
+  public async destroy (): Promise<void> {
+    for (const worker of this.workers) {
+      await this.destroyWorker(worker)
+    }
+  }
+
+  protected abstract destroyWorker (worker: Worker): void | Promise<void>
+
+  protected abstract sendToWorker (
+    worker: Worker,
+    message: MessageValue<Data>
+  ): void
+
+  /**
+   * Execute the task specified into the constructor with the data parameter.
+   *
+   * @param data The input for the task specified.
+   * @returns Promise that is resolved when the task is done.
+   */
+  public execute (data: Data): Promise<Response> {
+    // configure worker to handle message with the specified task
+    const worker = this.chooseWorker()
+    const previousWorkerIndex = this.tasks.get(worker)
+    if (previousWorkerIndex !== undefined) {
+      this.tasks.set(worker, previousWorkerIndex + 1)
+    } else {
+      throw Error('Worker could not be found in tasks map')
+    }
+    const id = ++this.id
+    const res = this.internalExecute(worker, id)
+    this.sendToWorker(worker, { data: data || ({} as Data), id: id })
+    return res
+  }
+
+  protected abstract registerWorkerMessageListener (
+    port: Worker,
+    listener: (message: MessageValue<Response>) => void
+  ): void
+
+  protected abstract unregisterWorkerMessageListener (
+    port: Worker,
+    listener: (message: MessageValue<Response>) => void
+  ): void
+
+  protected internalExecute (worker: Worker, id: number): Promise<Response> {
+    return new Promise((resolve, reject) => {
+      const listener: (message: MessageValue<Response>) => void = message => {
+        if (message.id === id) {
+          this.unregisterWorkerMessageListener(worker, listener)
+          const previousWorkerIndex = this.tasks.get(worker)
+          if (previousWorkerIndex !== undefined) {
+            this.tasks.set(worker, previousWorkerIndex + 1)
+          } else {
+            throw Error('Worker could not be found in tasks map')
+          }
+          if (message.error) reject(message.error)
+          else resolve(message.data as Response)
+        }
+      }
+      this.registerWorkerMessageListener(worker, listener)
+    })
+  }
+
+  protected chooseWorker (): Worker {
```

Comment:
That one is targeted to become a strategy (a callback with a defined signature) to be choose at context call ;)

Reply:
Yeah... I already tried to tackle that, BUT in dynamic-*-pool it uses `super.chooseWorker()` 😬

```ts
type IDynamicPool<Worker extends IWorker, Data, Response> = (
  | DynamicClusterPool<Data, Response>
  | DynamicThreadPool<Data, Response>
) &
  AbstractPool<Worker, Data, Response>

export function dynamicWorkerChooser<
  Worker extends IWorker,
  Data,
  Response,
  DynamicPool extends IDynamicPool<Worker, Data, Response>
> (this: DynamicPool, superChooseWorker: () => Worker): Worker {
  // ...
```

And then it could be called e.g. like this:
```ts
return dynamicWorkerChooser.bind(this, () => super.chooseWorker())
```

But when accessing functions like `internalNewWorker` or `sendToWorker` in `this(: DynamicPool)`, it cannot access them, because it tells me that `this` has no method defined like the ones :confused:

Reply:
@jerome-benoit  do you have any further suggestion on this?

Reply:
@pioardi Let's move on with the discussion in #122 so we don't suffer in chaos :slightly_smiling_face: